src/olympia/amo/tests/__init__.py

```diff
@@ -969,34 +968,44 @@ def copy_file_to_temp(source):
         yield temp_filename
 
 
-# This sets up a module that we can patch dynamically with URLs.
-@override_settings(ROOT_URLCONF='olympia.amo.tests.dynamic_urls')
-class WithDynamicEndpoints(TestCase):
+class WithDynamicEndpointsMixin(object):
     """
-    Mixin to allow registration of ad-hoc views.
+    Mixin to allow registration of ad-hoc views. The class using it *must* be
+    decorated with:
+    @override_settings(ROOT_URLCONF='olympia.amo.tests.dynamic_urls')
     """
 
     def endpoint(self, view, url_regex=None):
         """
-        Register a view function or view class temporarily
-        as the handler for requests to /dynamic-endpoint
+        Register a view function or view class temporarily as the handler for
+        requests to /api/v4/dynamic-endpoint (We use /api/v4/ to make sure not
+        to be affected by the locale & app redirection middleware.)
         """
-        url_regex = url_regex or r'^dynamic-endpoint$'
-        try:
-            is_class = issubclass(view, APIView)
-        except TypeError:
-            is_class = False
-        if is_class:
+        url_regex = url_regex or r'^api/v4/dynamic-endpoint$'
+        if hasattr(view, 'as_view'):
             view = view.as_view()
 
-        dynamic_urls.urlpatterns = [django_urls.url(url_regex, view)]
+        dynamic_urls.urlpatterns = [django_urls.url(
+            url_regex, view, name='test-dynamic-endpoint')
+        ]
 
         self.addCleanup(self._clean_up_dynamic_urls)
 
     def _clean_up_dynamic_urls(self):
         dynamic_urls.urlpatterns = []
 
 
+@override_settings(ROOT_URLCONF='olympia.amo.tests.dynamic_urls')
+class WithDynamicEndpoints(WithDynamicEndpointsMixin, TestCase):
+    pass
+
+
+@override_settings(ROOT_URLCONF='olympia.amo.tests.dynamic_urls')
+class WithDynamicEndpointsAndTransactions(
```

Comment:
Not ideal, but we can't decorate a mixin with `override_settings`, so we have to have concrete classes, and we definitely need the class to be a `TransactionTestCase` - it wasn't previously and that was completely messing up the test.

Reply:
Still much better than messed up tests